src/iqm/iqm_client/authentication.py

```diff
@@ -0,0 +1,308 @@
+        exp_time = int(json.loads(b64decode(body)).get('exp', '0'))
+        return max(0, exp_time - int(time.time()))
+
+    def __init__(
+        self,
+        token: Optional[str] = None,
+        tokens_file: Optional[str] = None,
+        auth_server_url: Optional[str] = None,
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+    ):
+        """
+        Initializes and configures the TokenManager.
+
+        Keyword Args:
+            token (str): Optional long-lived token in plain text format.
+                If ``token`` is given no other user authentication parameters should be given.
+            tokens_file (str): Optional path to a tokens file used for authentication.
+                If ``tokens_file`` is given no other user authentication parameters should be given.
+            auth_server_url (str): Optional base URL of the authentication server.
+                If ``auth_server_url`` is given also ``username`` and ``password`` must be given.
+            username (str): Optional username to log in to authentication server.
+            password (str): Optional password to log in to authentication server.
+
+        Alternatively, the user authentication related keyword arguments can also be given in
+        environment variables ``IQM_TOKEN``, ``IQM_TOKENS_FILE``, ``IQM_AUTH_SERVER``,
+        ``IQM_AUTH_USERNAME`` and ``IQM_AUTH_PASSWORD``. All parameters must be given either
+        as keyword arguments or as environment variables.
+        """
+
+        def _format_names(variable_names: list[str]) -> str:
+            """Format a list of variable names"""
+            return ', '.join(f'"{name}"' for name in variable_names)
+
+        auth_parameters: dict[str, str] = {}
+
+        init_parameters = {
+            'token': token,
+            'tokens_file': tokens_file,
+            'auth_server_url': auth_server_url,
+            'username': username,
+            'password': password,
+        }
+        init_params_given = [key for key, value in init_parameters.items() if value]
+
+        env_variables = {
+            'token': 'IQM_TOKEN',
+            'tokens_file': 'IQM_TOKENS_FILE',
+            'auth_server_url': 'IQM_AUTH_SERVER',
+            'username': 'IQM_AUTH_USERNAME',
+            'password': 'IQM_AUTH_PASSWORD',
+        }
+        env_vars_given = [value for key, value in env_variables.items() if os.environ.get(value)]
+
+        if init_params_given and env_vars_given:
+            raise ClientConfigurationError(
+                'Conflicting authentication parameters given: '
+                + f'keyword args {_format_names(init_params_given)} '
+                + f'and environment variables {_format_names(env_vars_given)}.'
+            )
+
+        if env_vars_given:
+            env_parameters = {key: os.environ.get(value) for key, value in env_variables.items()}
+            auth_parameters = {key: str(value) for key, value in env_parameters.items() if value}
```

Comment:
```suggestion
            auth_parameters = {key: value for key, envvar in env_variables.items() if (value := os.environ.get(envvar)}
```
os.environ should always be dict[str,str]

Reply:
fixed